packages/core/src/projections/lock.ts

```diff
@@ -0,0 +1,66 @@
+import { Mapping } from "@ddd-ts/shape";
+
+/**
+ * More advanced lock system, to handle maximum parallel processing concurrency restrictions
+ */
+export class Lock extends Mapping([String, String]) {
+  get(k: string) {
+    return this.value[k];
+  }
+
+  has(k: string) {
+    return this.value[k] !== undefined;
+  }
+
+  restrains(other: Lock, strict = true): boolean {
+    const { onlyLeft, onlyRight, both, bothHaveSameValues } = Lock.differences(
+      this,
+      other,
+    );
+
+    if (onlyLeft.length === 0) {
+      if (onlyRight.length === 0) {
+        if (bothHaveSameValues) {
+          return strict;
+        }
+      }
+    }
+
+    if (onlyLeft.length > 0) {
+      if (onlyRight.length === 0) {
+        if (bothHaveSameValues) {
+          return false;
+        }
+      }
+    }
+
+    if (onlyLeft.length === 0) {
+      if (onlyRight.length > 0) {
+        if (bothHaveSameValues) {
+          return true;
+        }
+      }
+    }
+
+    return false;
+  }
+
+  static allKeys(left: Lock, right: Lock) {
+    return new Set([...Object.keys(left.value), ...Object.keys(right.value)]);
+  }
+
+  static differences(left: Lock, right: Lock) {
```

Comment:
nitpick but you can improve perf by doing only one run : 

  static differences(left: Lock, right: Lock) {
    const onlyLeft: string[] = [];
    const onlyRight: string[] = [];
    const both: string[] = [];
    let bothHaveSameValues = true;

    for (const key of Lock.allKeys(left, right)) {
      const leftHas = left.has(key);
      const rightHas = right.has(key);

      if (leftHas && rightHas) {
        both.push(key);
        if (bothHaveSameValues && left.get(key) !== right.get(key)) {
          bothHaveSameValues = false;
        }
      } else if (leftHas) {
        onlyLeft.push(key);
      } else {
        onlyRight.push(key);
      }
    }

    return { onlyLeft, onlyRight, both, bothHaveSameValues };
  }